File: agent/native-cpp/src/prompt/conversation_context_prompt.cpp
#include "agent/prompt/conversation_context_prompt.h"

#include <array>
#include <sstream>

namespace atrium::agent {

namespace {

struct SectionSpec {
    ConversationContextEntryKind kind;
    const char *title;
};

constexpr std::array<SectionSpec, 8> kSections{{
    {ConversationContextEntryKind::Goal, "Current goals"},
    {ConversationContextEntryKind::Constraint, "Active constraints"},
    {ConversationContextEntryKind::Decision, "Active decisions"},
    {ConversationContextEntryKind::RejectedOption, "Rejected options"},
    {ConversationContextEntryKind::OpenQuestion, "Open questions"},
    {ConversationContextEntryKind::Risk, "Risks and conflicts"},
    {ConversationContextEntryKind::KeyFact, "Key facts"},
    {ConversationContextEntryKind::ProgressNote, "Progress notes"},
}};

void appendSources(std::ostringstream &out, const ConversationContextEntry &entry, bool include_source_ids) {
    if (!include_source_ids || entry.sources.empty()) {
        return;
    }

    out << " [sources:";
    for (std::size_t i = 0; i < entry.sources.size(); ++i) {
        if (i > 0) {
            out << ",";
        }
        out << " #" << entry.sources[i].message_id;
    }
    out << "]";
}

} // namespace

void appendConversationContextToPrompt(PromptPlan &plan, const ConversationContextState &state, ConversationContextPromptOptions options) {
    if (state.empty()) {
        return;
    }

    plan.addSystem("conversation-context", buildConversationContextBlock(state, options));
}

std::string buildConversationContextBlock(const ConversationContextState &state, ConversationContextPromptOptions options) {
    std::ostringstream out;

    out << "Conversation context state for conversation #" << state.conversation_id << ".\n";
    out << "Use this as persistent discussion state. Respect active constraints and decisions. Do not repeat rejected options unless the user reopens them.\n";

    if (state.last_summarized_message_id != 0) {
        out << "Summary covers messages through #" << state.last_summarized_message_id << ".\n";
    }

    if (!state.summary.empty()) {
        out << "\nSummary:\n";
        out << state.summary << "\n";
    }

    for (const auto &section : kSections) {
        auto entries = entriesByKind(state, section.kind, options.include_inactive_entries);
        if (entries.empty()) {
            continue;
        }

        out << "\n" << section.title << ":\n";
        std::size_t emitted = 0;
        for (const auto &entry : entries) {
            if (options.max_entries_per_kind > 0 && emitted >= options.max_entries_per_kind) {
                break;
            }
            out << "- ";
            if (entry.status != ConversationContextEntryStatus::Active) {
                out << "[" << toString(entry.status) << "] ";
            }
            out << entry.content;
            appendSources(out, entry, options.include_source_ids);
            out << "\n";
            ++emitted;
        }
    }

    return out.str();
}

} // namespace atrium::agent

File: agent/src/prompt/conversationContextPrompt.ts
import { entriesByKind } from "../context/conversationContext.ts";
import { ConversationContextEntryKind, ConversationContextEntryStatus } from "../context/conversationContext.ts";
import type { ContextProvenance, ConversationContextEntry, ConversationContextState } from "../context/conversationContext.ts";
import { promptSafeLabel } from "../core/agentTypes.ts";
import { PromptPlan, PromptSegment } from "./promptPlan.ts";

export interface ConversationContextPromptOptions {
  readonly includeSourceIds: boolean;
  readonly includeInactiveEntries: boolean;
  readonly maxEntriesPerKind: number;
}

export const DEFAULT_CONVERSATION_CONTEXT_PROMPT_OPTIONS: ConversationContextPromptOptions = {
  includeSourceIds: true,
  includeInactiveEntries: false,
  maxEntriesPerKind: 8,
};

const SECTIONS: readonly { kind: ConversationContextEntryKind; title: string }[] = [
  { kind: ConversationContextEntryKind.Goal, title: "Current goals" },
  { kind: ConversationContextEntryKind.Constraint, title: "Confirmed constraints" },
  { kind: ConversationContextEntryKind.Decision, title: "Confirmed decisions" },
  { kind: ConversationContextEntryKind.CurrentDirection, title: "Confirmed current direction" },
  { kind: ConversationContextEntryKind.RejectedOption, title: "Rejected options" },
  { kind: ConversationContextEntryKind.OpenQuestion, title: "Open questions" },
  { kind: ConversationContextEntryKind.Risk, title: "Risks" },
  { kind: ConversationContextEntryKind.KeyFact, title: "Key facts" },
  { kind: ConversationContextEntryKind.ProgressNote, title: "Progress notes" },
];

export function appendConversationContextToPrompt(
  plan: PromptPlan,
  state: ConversationContextState,
  options: ConversationContextPromptOptions = DEFAULT_CONVERSATION_CONTEXT_PROMPT_OPTIONS,
): void {
  plan.addSystem(PromptSegment.ConfirmedWhiteboard, "conversation-context", buildConversationContextBlock(state, options));
}

export function buildConversationContextBlock(
  state: ConversationContextState,
  options: ConversationContextPromptOptions = DEFAULT_CONVERSATION_CONTEXT_PROMPT_OPTIONS,
): string {
  const lines = [
    `Confirmed conversation whiteboard for conversation #${promptSafeLabel(state.conversationId)}.`,
    `Context version: ${promptSafeLabel(state.contextVersion)}.`,
    `Phase: ${state.phase}.`,
    "Only confirmed material appears here. Treat proposals and drafts elsewhere as unconfirmed.",
  ];

  if (state.lastSummarizedMessageId !== "0") {
    lines.push(`Summary covers messages through #${promptSafeLabel(state.lastSummarizedMessageId)}.`);
  }
  if (state.summary.length > 0) {
    lines.push("", "Summary:", state.summary);
  }

  for (const section of SECTIONS) {
    const entries = entriesByKind(state, section.kind, options.includeInactiveEntries);
    if (entries.length === 0) {
      continue;
    }
    lines.push("", `${section.title}:`);
    const limit = options.maxEntriesPerKind > 0 ? options.maxEntriesPerKind : entries.length;
    for (const entry of entries.slice(0, limit)) {
      const status = entry.status === ConversationContextEntryStatus.Active ? "" : `[${entry.status}] `;
      lines.push(`- ${status}${formatEntryContent(entry)}${formatSources(entry.sources, options.includeSourceIds)}`);
    }
  }

  if (state.summary.length === 0 && state.entries.length === 0) {
    lines.push("", "No confirmed whiteboard content yet.");
  }
  return lines.join("\n");
}

function formatEntryContent(entry: ConversationContextEntry): string {
  if (entry.kind !== ConversationContextEntryKind.RejectedOption || !entry.rejectedOption) {
    return entry.content;
  }
  return `option: ${entry.rejectedOption.option} | reason: ${entry.rejectedOption.reason} | premise: ${entry.rejectedOption.premise}`;
}

function formatSources(sources: readonly ContextProvenance[], includeSourceIds: boolean): string {
  if (!includeSourceIds) {
    return "";
  }
  return ` [sources:${sources.map(formatSource).join(",")}]`;
}

function formatSource(source: ContextProvenance): string {
  if (source.type === "message") {
    return ` #${promptSafeLabel(source.messageId)}:${source.status}`;
  }
  return ` ${source.type}:${promptSafeLabel(source.provenanceId)}`;
}
